### ISSUE TITLE
Expose Microsoft.Framework.Runtime.Hosting to .NET 4.5

### ISSUE THREAD

**Issue by mkosieradzki** [aspnet/dnx#1431]
Scenario: I would like to use KRE based apps as extensibility point for my system utilizing .NET Framework 4.5+.

There are multiple loaders for native environment, but I am unable to find proper, developer-friendly :) loader for the full .NET Framework.
Of course there is Helios loader - but it is quite tied to the IIS.
kre.clr.managed.dll alone - is not really well documented and is not the most handy tool to use :) it also requires using separated appdomain.

It would be great to be able to use Microsoft.Framework.Runtime.DefaultHost class from .NET Framework application directly, but not sure if this is even possible due to the new dependency loading and assembly neutral types mechanisms.

I would be also grateful for some pointers on preferred way of accomplishing my scenario.

**Comment by davidfowl** [comment]
This isn't something we're currently aiming to make a first class API in the short term. We are working on a rewrite that will allow this https://github.com/aspnet/dnx/tree/dev/src/Microsoft.Framework.Runtime.Hosting, but the first goal is to making it efficient rather than making it a nice public API.

What behaviors do you want to use exactly?

**Comment by mkosieradzki** [comment]
Thank you very much so quick answer.

I wanted to use automatic Roslyn Loaders, Nuget Loaders and proper assembly neutral types support and project.json. But it looks like I have it up and running even without separate AppDomain (a lot of source code reading behind me). 

The problem I faced was that assemblies for neutral types must be extracted to a directory where fusion can find them, while I was trying to load them from their in-mem binary images from Resources - kinda dumb of me.

To sum up - got it running using Microsoft.Framework.Runtime.DefaultHost - but this is a simple prototype. I need to do a lot of testing now in more complicated scenarios. I hope this is a supported (or will be supported scenario) and I am not doing things completely against the design.

btw. any chance Microsoft.Framework.Runtime will be added back to .NET Framework v4.5 nuget package like it was in beta-1 version? It's currently marked as ASP.NET 5/ASP .NET CORE only.

**Comment by davidfowl** [comment]
Microsoft.Framework.Runtime is going to be phased out in favor of the other dll I posted. Assembly neutral interfaces were also removed. As for adding the net45 target, that's possible we'll see if it makes sense.

**Comment by mkosieradzki** [comment]
I think I understand why :) - they are kinda nightmarish to use while bridging .net 4.5 with asp net 5 (from the .NET side). But they sounded like a great idea in asp.net 5 alone. OK. So if there are so important changes in-progress/ahead I will limit my development against those assemblies and try to keep up with the latest developments including the assembly you have mentioned.

Thank you very much for all the priceless informations.

**Comment by mkosieradzki** [comment]
After clarifications in:
https://github.com/aspnet/HttpAbstractions/issues/301 and https://github.com/aspnet/Home/issues/532

Please expose Microsoft.Framework.Runtime.Hosting to .NET 4.5.

**Comment by LodewijkSioen** [comment]
+1

I help run a large business application that is mostly Webforms. We were planning to start using MVC alongside Webforms (like we already do with SignalR), but it would be painful if we had to stay in MVC5 land indefinitely.
There was all this brouhaha around 'One Asp.net' and 'lego pieces that fit together', but this seems to be an entirely new baseplate (to keep the Lego analogy).

**Comment by mkosieradzki** [comment]
@davidfowl What has happened to  https://github.com/aspnet/dnx/tree/dev/src/Microsoft.Framework.Runtime.Hosting you had mentioned before? It has dissapeared from the dev branch some time ago.

**Comment by davidfowl** [comment]
@mkosieradzki We're refactoring the existing logic instead of forking and trying to rewrite.

**Comment by davidfowl** [comment]
@LodewijkSioen This is totally a new baseplate.

**Comment by mkosieradzki** [comment]
@davidfowl Thanks. So just to sum up. Microsoft.Framework.Runtime.Hosting is going to stay or will be provided in another way. It's not working yet and that's why it is not exposed in nightly feed: https://www.myget.org/gallery/aspnetvnext. It is planned to be exposed to .NET 4.5 and it is going to be the main bridge between .NET 4.5 and ASP.NET5 allowing to pass data in-process.

If I want to host ASP.NET 5 web application inside my .NET 4.5 application at RTM/RTW time, am I rather supposed to wait for this hosting facilities? Or should I "steal" building blocks from DNX451 (as they are not marked for net45) and build my own Microsoft.Framework.Runtime.Hosting-like implementation using Loaders and Microsoft.Framework.Runtime.Interfaces. Both are perfectly OK, but don't want to reinvent the wheel.

My plan was quite similiar to what you have suggested in aspnet/Hosting#291 (you have clarified a lot of plans there!): to create OWIN adapter for ASP.NET 5 and marshal OwinContext to ASP.NET5 request context.
BTW: For me it's quite convenient as I don't intend to use the default HTTP OwinContext but own OwinContext so I don't need any of facilities mentioned in aspnet/Hosting#291.

**Comment by mkosieradzki** [comment]
I am dropping this requirement and moving from cross-appdomain in-process hosting to separate process. It is less efficient but much more scalable.

**Comment by davidfowl** [comment]
@mkosieradzki Are you still interested in doing any of this work? I think it would be interesting as a community experiment to produce an implementation of https://github.com/aspnet/dnx/tree/dev/src/Microsoft.Dnx.Runtime.Abstractions and https://github.com/aspnet/dnx/tree/dev/src/Microsoft.Dnx.Compilation.Abstractions on vanilla .NET.

**Comment by mkosieradzki** [comment]
@davidfowl From my point of view it is still worth consideration as I haven't implemented the separate process hosting stuff yet. It's in the design phase and I can still reconsider going back to in-process runtime rehosting approach. I can see performance benefits coming from hosting DNX in vanilla .NET as well as I can see benefits allowing people to integrate ASP.NET 5 and ASP.NET 4.6 application in single process what can be huge in terms in step-by-step porting of existing ASP.NET 4.6 applications to the new leaner runtime.

I can see of course a lot of benefits in scenarios like mine: multiple small .NET 5 sub-apps hosted inside ASP.NET 4.6 app.

I am currently a bit swamped at my main projects, but I can try to find some time later this month, as long as it is not UNSUPPORTED by design. I mean: I don't want to invest time and effort into creating something that is not feasible by design. Also it would be great to know that if project succeeds, it would be integrated and maintained along with DNX. Implementation is one thing, maintenance completely different.

I went through code and see you guys did serious refactoring in the hosting area. I think this is still a great idea and now I understand why it was too early to do it before.

**Comment by MJGraf** [comment]
It would really be great if someone could provide a possibility to host ASP.Net 5 in vanilla .Net.

I fully understand and appreciate that DNX has many advantages over vanilla .Net, but in particular for already existing applications it may be hard to switch the whole solution to DNX at once. I could well imagine that many people (like me) would first want to try ASP.NET 5 as a "future-proof Katana replacement" without having to deal with DNX and maybe later on make the switch to DNX.
So having a possibility to host ASP.NET 5 in a vanilla .Net project may materially increase the user base of ASP.NET 5 (even if it was considered a non-first-class way of hosting).

I would love to help in making this happen, but after having a  look at the two ...Abstractions directories above, I am completely lost and it seems this task is far beyond my skill level. I suppose that the entry point should be WebHostBuilder, which requires particularly an IServiceProvider. But which interfaces need to be implemented apart from that is not clear to me...

**Comment by mkosieradzki** [comment]
@MJGraf If you want to work on this together it would be great to have you onboard.
My proposition is as follows: 
1. Let me create a prototype upcoming week (I really hope I can find some time for this project) in a forked/new repository on github. I had some attempts on this and even half-working prototypes in earlier DNX builds.
2. Let's then implement all the interfaces/services required to run:
a) Simple console application
b) Simple IApplicationBuilder-based hello world http server (from source code and pre-built app)
c) Example MVC6 applications
d) Other middlewares: which ones?
e) Larger MVC6 projects
f)  ... whatever else we may need to consider implementation complete.
3. Let's figure out how many changes in core repositories we need to achieve appropriate (efficient) code reuse and investigate whether our implementation can be co-maintained with DNX or not.

To be honest I cannot predict point 3 at this moment, but I have some ideas how to handle it.

**Comment by MJGraf** [comment]
@mkosieradzki I'm happy to lend a hand to the extent I can understand it.
My primary usecase currently is that I implemented a very simple Owin/Katana based webserver for MediaPortal 2 (https://github.com/MediaPortal/MediaPortal-2/commits/FEAT_MP2-437_OwinServer) some time ago. It was more a proof of concept and was meant to replace the very old (and no longer maintained) HttpServer we currently use for streaming media files via http and our implementation of a UPnP server. I would therefore in particular need the StaticFiles Middleware, but WebAPI will be needed as well later on. Unfortunately I have no knowledge at all about MVC.
MediaPortal 2 has its own quite sophisticated plugin system, where every plugin (in a standard installation maybe 20 or so) consists of a separate dll in its own directory (together with a descriptive plugin.xml file). Every such plugin is therefore an own .csproj. The server application can currently be run as a Windows Service (and for debugging purposes also as console application). Switching all these projects to dnx would be quite a challenge (apart from the fact that I haven't found out, yet, how DNX plays together with WiX, which we use for deployment purposes).
I was planning to research whether I can just switch our "main" projects to DNX and let most of the plugin projects (except for the web server plugin) vanilla .net 451 projects. It seems that running a dnx project as Windows Service is possible (http://taskmatics.com/blog/run-dnx-applications-windows-service/), but what seems to be difficult for me is that dnx projects may reference vanilla .net projects, but not (yet) vice versa.
That's where I am now... And the easiest solution for me would be (at least for now) to just host ASP.NET 5 in a vanilla .net project.
Just let me know when you have your prototype pushed and I will have a look at it!
Thanks!

**Comment by mkosieradzki** [comment]
I found couple of hours today to do some research.
@davidfowl From my short research it makes no sense to reimplement vanilla host as you propose (I might be very wrong).

I started from adding additional target: net45 to all hosting projects in DNX and ended up in completely reproducing dnx451. I have created pure vanilla host that uses three assemblies from DNX: Microsoft.Dnx.Host, Microsoft.Dnx.Loader, Microsoft.Dnx.Runtime.Abstraction (dnx451 target is absolutely good enough for net45 prototype).

After some experiments (reverse engineering the way command line arguments are handled throughout multiple layers of DNX), I have succeeded to run hello world, and to find how to pass and configure core parameters to the hosting app domain. There are well documented Environment variables and so on. I expect to have this working.

Running precompiled Hello World is not much. But next step is to run Microsoft.Dnx.ApplicationHost which can do all the magic: Loading shared assemblies from many locations, compiling sources etc. I predict it highly feasible task.

Further steps can include cross-app-domain remoting which shouldn't be difficult to add, although it might require some alteration in startup sequence.

@MJGraf Have a look at: https://github.com/mkosieradzki/dnx . It's not very impressive, but works with precompiled console application. I assume that approach I am using will allow to reuse nearly all DNX so code it should be highly compatible.

**Comment by mkosieradzki** [comment]
Small update: I have successfully managed to run HelloWorld sample from source code / nuget using standard .NET 4.5 application using all the magic of Microsoft.Dnx.ApplicationHost. I have uploaded updated code.

@MJGraf I will have yet another extremely busy week and will have no time to work on this project. Assuming I have console applications up and running, I see no reasons now to have any problems running even most complex ASP.NET MVC6 apps. There might be some work to be done in integrating HttpListener (or something :) - I don't remember current name of this component). But it should be quite easy from where I ended.

**Comment by mkosieradzki** [comment]
I had some really great progress with my research so I decided to complete this today.

@davidfowl I will need to add net45 targets to three projects generating outputs:
Microsoft.Dnx.Runtime.Abstractions - you probably wanted to do this anyway.
Microsoft.Dnx.Loader - it's actually a small assembly (3 files) - so alternatively we can just reuse source code, but I see no good reason to not add net45
Microsoft.Dnx.Host - together with it's source dependencies: (Microsoft.Framework.CommandLineUtils.Sources, Microsoft.Dnx.Runtime.Internals, Microsoft.Dnx.Runtime.Sources. -if you think it's better to reuse source code... But in my opinion this is so small shared surface it can be safely exposed to net45.

I will also need to alter all ifdefs in those files (DNX451 -> DNX451 || NET45) for obvious reasons.

After those modifications running DNX apps, Microsoft.Dnx.ApplicationHost is quite easy and does not require any retargeting.

Vanilla host creation will look as follows: include nugget packages for the following three components: 1) Use Microsoft.Dnx.Host.RuntimeBootstrapper (after settings environment)
OR
2) Use Microsoft.Dnx.Host.Bootstrapper - just by passing appropriate arguments
OR
3) Implement own bootstrapper using code copied from Microsoft.Dnx.Host + Microsoft.Dnx.Runtime.*

I have created my own bootstrapper to allow passing services to the hosted application.

I will create a pull request with net45 targets.
And later I will create a second pull request with conditional NET45 feature allowing to pass services to the bootstrapper.

This approach will allow to achieve best maintainability.

**Comment by davidfowl** [comment]
I should probably have warned you before sending this PR that we're not ready to take this change yet. The API in the dnx is also no stable and we plan to internalize most of it since we don't really want people to depend on it directly. It's great that you got it working and we do want to make the APIs available more broadly but this isn't a v1 thing. We may add more targets but I recommend using your fork until we figure out exactly what set of APIs we want to support.

**Comment by mkosieradzki** [comment]
@davidfowl Don't worry David. The most effort went to the research and I can build up a fully function working host just by "stealing" 3 assemblies from DNX451 build. The only part of code that would be really helpful to have inside DNX codebase is an alternative bootstrapper that allows passing services from host. (I will create a feature request and can also create a PR if you like, but that's definitely not a rocket science).

BTW. regarding your previous post: .NET 4.5 version of Microsoft.Dnx.Compilation.Abstractions is definitely not necessary to have it all up and running while Microsoft.Dnx.Runtime.Abstractions is critical (but it is binary equivalent to the DNX451 version so no that's not a blocking problem at the moment).

Really great stuff about all this things is how thin the external hosting layer is and how much is done by xplat Microsoft.Dnx.ApplicationHost - I really love this design. Even if I stick to my fork I can have it compatible with any runtime as long as they use the same version of Microsoft.Dnx.Runtime.Abstractions.

**Comment by davidfowl** [comment]
@mkosieradzki Great work! 

> BTW. regarding your previous post: .NET 4.5 version of Microsoft.Dnx.Compilation.Abstractions is definitely not necessary to have it all up and running while Microsoft.Dnx.Runtime.Abstractions is critical (but it is binary equivalent to the DNX451 version so no that's not a blocking problem at the moment).

Microsoft.Dnx.Compilation.Abstractions is required for things like MVC to run. Razor compilation relies on this today.

> Really great stuff about all this things is how thin the external hosting layer is and how much is done by xplat Microsoft.Dnx.ApplicationHost - I really love this design. Even if I stick to my fork I can have it compatible with any runtime as long as they use the same version of Microsoft.Dnx.Runtime.Abstractions.

Yep! Thanks! :smile:

**Comment by muratg** [comment]
Putting this one in the Backlog.

**Comment by davidfowl** [comment]
With the move to the new CLI we've made libraries for the brains of DNX (not the execution part as yet). See https://github.com/dotnet/cli for more information.
